lib/hooks.dart

```diff
@@ -0,0 +1,97 @@
+@JS()
+library hooks;
+
+import 'package:js/js.dart';
+import 'package:react/react.dart';
+
+@JS()
+abstract class React {
+  external static List<dynamic> useState(dynamic value);
+}
+
+/// The return value of [useState].
+///
+/// The current value of the state is available via [value] and
+/// functions to update it are available via [set] and [setTx].
+///
+/// Note are two rules for using Hooks (<https://reactjs.org/docs/hooks-rules.html>):
+///
+/// * Only call Hooks at the top level.
+/// * Only call Hooks from inside a [DartFunctionComponent].
+///
+/// Learn more: <https://reactjs.org/docs/hooks-state.html>.
+class StateHook<T> {
+  /// The first item of the pair returned by [React.useState].
+  T _value;
+
+  /// The second item in the pair returned by [React.useState].
+  void Function(dynamic) _setValue;
+
+  StateHook(T initialValue) {
+    final result = React.useState(initialValue);
+    _value = result[0];
+    _setValue = result[1];
+  }
+
+  StateHook.init(T init()) {
+    final result = React.useState(allowInterop(init));
+    _value = result[0];
+    _setValue = result[1];
+  }
+
+  /// The current value of the state.
+  ///
+  /// See: <https://reactjs.org/docs/hooks-reference.html#usestate>.
+  T get value => _value;
+
+  /// Updates [value] to [newValue].
+  ///
+  /// See: <https://reactjs.org/docs/hooks-state.html#updating-state>.
+  void set(T newValue) => _setValue(newValue);
+
+  /// Updates [value] to the return value of [computeNewValue].
+  ///
+  /// See: <https://reactjs.org/docs/hooks-reference.html#functional-updates>.
+  void setTx(T computeNewValue(T oldValue)) => _setValue(allowInterop(computeNewValue));
```

Comment:
It was mean to be "transactional", but it was clear when I had `setStateTx` that that wasn't clear haha...

Perhaps `setWithUpdater` would be better? Or even just `update`? But with the latter it might not be the immediately obvious the difference between `set` and `update`.

Thoughts? I think that for these kinds of naming decisions, we should gather input from more rather than fewer people. @joebingham-wk @kealjones-wk

Reply:
i think `setWithUpdater` makes a lot of sense as it correlates to component2's new `setStateWithUpdater`, also we never made a codemod for that did we .... should we?

Reply:
I like `setWithUpdater`